### PR TITLE
Remove jQuery class from the repository topic box

### PR DESCRIPTION
**Pull request by yardenshoham**
- Switched from jQuery class functions to plain JavaScript `classList`
- Tested the repository topic box functionality and it works as before
- Various fixes by wxiaoguang

### CODE REVIEW

**Review by delvh — APPROVED** [review]

File: web_src/js/features/repo-home.js [hunk 1]
```diff
@@ -152,12 +155,14 @@ export function initRepoTopicBar() {
   });
 
   $.fn.form.settings.rules.validateTopic = function (_values, regExp) {
-    const $topics = $topicDropdown.children('a.ui.label');
-    const status = !$topics.length || $topics.last()[0].getAttribute('data-value').match(regExp);
+    const topics = topicDropdown.querySelectorAll(':scope > a.ui.label');
+    const lastTopic = topics[topics.length - 1];
+    const status = !topics.length || lastTopic.getAttribute('data-value').match(regExp);
```
Comment (silverwind) [inline web_src/js/features/repo-home.js]:
```suggestion
    const status = !topics.length || (topics[topics.length - 1]).getAttribute('data-value').match(regExp);
```

Slightly better as this should avoid accessing the array at index `-1`, but I still find this code hard to follow.

File: web_src/js/features/repo-home.js [hunk 2]
```diff
@@ -152,12 +155,14 @@ export function initRepoTopicBar() {
...
     if (!status) {
-      $topics.last().removeClass('green').addClass('red');
+      lastTopic.classList.remove('green');
+      lastTopic.classList.add('red');
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]: It doesn't seem right. If `!status`, it might also means "no last topic"

Comment (yardenshoham) [reply web_src/js/features/repo-home.js]: https://github.com/go-gitea/gitea/pull/30191/commits/76d91b0aa122d6c1b49f54f1323128f26de817db

File: web_src/js/features/repo-home.js [hunk 3]
```diff
@@ -152,12 +155,13 @@ export function initRepoTopicBar() {
...
   $.fn.form.settings.rules.validateTopic = function (_values, regExp) {
-    const $topics = $topicDropdown.children('a.ui.label');
-    const status = !$topics.length || $topics.last()[0].getAttribute('data-value').match(regExp);
-    if (!status) {
-      $topics.last().removeClass('green').addClass('red');
+    const topics = topicDropdown.querySelectorAll(':scope > a.ui.label');
+    const status = !topics.length || (topics[topics.length - 1]).getAttribute('data-value').match(regExp);
+    if (!status && topics[topics.length - 1]) {
+      (topics[topics.length - 1]).classList.remove('green');
+      (topics[topics.length - 1]).classList.add('red');
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]: TBH I am not a fan of such refactoring style. As a refactoring PR, if it could be done with simple search&replace, then it could keep the old logic, and some tests could be skipped.

However, since it has been rewritten a lot and needs to be fully tested, I believe it's better to rewrite the code to a proper style.


```suggestion
    const topics = topicDropdown.querySelectorAll('a.ui.label');
    const lastTopic = topics[topics.length - 1];
    const isLastTopicValid = lastTopic?.getAttribute('data-value').match(regExp);
    if (lastTopic && !isLastTopicValid) {
      lastTopic.classList.remove('green');
      lastTopic.classList.add('red');
```

File: web_src/js/features/repo-home.js [hunk 4]
```diff
@@ -23,7 +25,7 @@ export function initRepoTopicBar() {
   mgrBtn.addEventListener('click', () => {
     hideElem(viewDiv);
     showElem(editDiv);
-    $topicDropdownSearch.trigger('focus');
+    topicDropdownSearch?.focus();
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]: It seems that the `topicDropdownSearch` should always be there.

```suggestion
    topicDropdownSearch.focus();
```

File: web_src/js/features/repo-home.js [hunk 5]
```diff
@@ -102,9 +105,9 @@ export function initRepoTopicBar() {
         const query = stripTags(this.urlData.query.trim());
         let found_query = false;
         const current_topics = [];
-        $topicDropdown.find('a.label.visible').each((_, el) => {
+        for (const el of topicDropdown.querySelectorAll(':scope > a.label.visible')) {
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]:
```suggestion
        for (const el of topicDropdown.querySelectorAll('a.label.visible')) {
```

`find` doesn't need scope.

File: web_src/js/features/repo-home.js [hunk 6]
```diff
@@ -152,12 +155,13 @@ export function initRepoTopicBar() {
...
-    if (!status) {
-      $topics.last().removeClass('green').addClass('red');
+    const topics = topicDropdown.querySelectorAll(':scope > a.ui.label');
+    const status = !topics.length || (topics[topics.length - 1]).getAttribute('data-value').match(regExp);
+    if (!status && topics[topics.length - 1]) {
+      (topics[topics.length - 1]).classList.remove('green');
+      (topics[topics.length - 1]).classList.add('red');
     }
-    return status && !$topicDropdown.children('a.ui.label.red').length;
+    return status && !topicDropdown.querySelectorAll(':scope > a.ui.label.red').length;
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]: Since the `red` class has been added to the invalid labels:

```suggestion
    return topicDropdown.querySelectorAll('a.ui.label.red').length === 0;
```

"no red label" means the topics are all valid.

File: web_src/js/features/repo-home.js [hunk 7]
```diff
@@ -64,10 +66,11 @@ export function initRepoTopicBar() {
         topicPrompts.formatPrompt = responseData.message;
 
         const {invalidTopics} = responseData;
-        const $topicLabels = $topicDropdown.children('a.ui.label');
+        const topicLabels = topicDropdown.querySelectorAll(':scope > a.ui.label');
```
Comment (wxiaoguang) [inline web_src/js/features/repo-home.js]: I think it could simply use `topicDropdown.querySelectorAll('a.ui.label')`, because other places just use it (without scope)

Comment (yardenshoham) [reply web_src/js/features/repo-home.js]: https://github.com/go-gitea/gitea/pull/30191/commits/2cc10ac988defedf20a5d86261ad7ebdb1019ce5

**Comment by wxiaoguang** [comment]
Made some more changes:

* The old UI is broken (not by this PR, but a regression). Fix it.

<details>

image

</details>

* Remove frontend validation, because the server side already does so.

<details>

image

</details>

**Comment by denyskon** [comment]
See #29849 for the discussion about replacement of fomantic dropdowns and other fomantic components

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
For the UI layout bug, I created a quick fix PR for 1.22:  Fix home topic edit form layout #30213 

So here are 2 choices for it:

1. Backport this PR
2. Apply that quick fix

File: web_src/js/utils/dom.js [hunk 8]
```diff
@@ -51,8 +51,22 @@ export function isElemHidden(el) {
...
+  }
+  return elems;
+}
+
+export function queryElemSiblings(el, selector = '*', fn = null) {
+  return applyElemsCallback(Array.from(el.parentNode.children).filter((child) => child !== el && child.matches(selector)), fn);
+}
+
+// it works like jQuery.children: only the direct children are selected
+export function queryElemChildren(parent, selector, fn = null) {
```
Comment (silverwind) [inline web_src/js/utils/dom.js]:
```suggestion
export function queryElemChildren(parent, selector = '*', fn = null) {
```

File: web_src/js/utils/dom.js [hunk 9]
```diff
@@ -51,8 +51,22 @@ export function isElemHidden(el) {
   return res[0];
 }
 
-export function queryElemSiblings(el, selector = '*') {
-  return Array.from(el.parentNode.children).filter((child) => child !== el && child.matches(selector));
+function applyElemsCallback(elems, fn) {
+  if (fn) {
+    for (const el of elems) {
+      fn.call(el, el); // call the function with the element as "this", and pass the element as the first argument
+    }
```
Comment (silverwind) [inline web_src/js/utils/dom.js:58]: I wouldn't encourage `this` usage for new code, so this should just be `fn(el)`.

**Review by silverwind — CHANGES_REQUESTED** [review]
See comments above

File: web_src/js/utils/dom.js (hunk 9 above)
Comment (wxiaoguang) [inline web_src/js/utils/dom.js]:
```suggestion
    for (const el of elems) {
      fn(el);
    }
```

`this` makes it easier to migrate existing jQuery code to the new framework.

But if you insist to remove `this` support, feel free to change it.

Comment (wxiaoguang) [reply web_src/js/utils/dom.js:58]: See https://github.com/go-gitea/gitea/pull/30191#pullrequestreview-1970253726

Comment (silverwind) [reply web_src/js/utils/dom.js]: Just use the first argument, all `this` usage in such functions should be forbidden in new code.

File: web_src/js/utils/dom.js [hunk 10]
```diff
@@ -51,8 +51,22 @@ export function isElemHidden(el) {
...
+function applyElemsCallback(elems, fn) {
+  if (fn) {
+    for (const el of elems) {
+      fn(el);
+    }
+  }
+  return elems;
+}
+
+export function queryElemSiblings(el, selector = '*', fn = null) {
```
Comment (silverwind) [inline web_src/js/utils/dom.js]:
```suggestion
export function queryElemSiblings(el, selector = '*', fn) {
```

File: web_src/js/utils/dom.js [hunk 11]
```diff
@@ -51,8 +51,22 @@ export function isElemHidden(el) {
...
+  return applyElemsCallback(Array.from(el.parentNode.children).filter((child) => child !== el && child.matches(selector)), fn);
+}
+
+// it works like jQuery.children: only the direct children are selected
+export function queryElemChildren(parent, selector = '*', fn = null) {
```
Comment (silverwind) [inline web_src/js/utils/dom.js]:
```suggestion
export function queryElemChildren(parent, selector = '*', fn) {
```

File: web_src/js/utils/dom.js (hunk 10 above)
Comment (silverwind) [reply web_src/js/utils/dom.js]: No reason to default to null when default `undefined` works just as well.

Comment (wxiaoguang) [reply web_src/js/utils/dom.js]: 1. In most languages (maybe all except JS), `foo(a, b=1, c)` is incorrect syntax, and it looks really ugly.
2. `null` is shorter than `undefined`, and in this case it just works well.

Feel free to change to what you prefer.

Comment (silverwind) [reply web_src/js/utils/dom.js]: `null` is problematic with default parameters because they only work with undefined.

More reasons to avoid null in https://github.com/sindresorhus/meta/discussions/7, I don't agree to remove null completely in code as I still like to use it in certain cases.

**Review by silverwind — APPROVED** [review]
